### DISCUSSION TITLE
First Public Developer Preview 0.3 Alpha1 Released & statement~

### DISCUSSION THREAD

**Discussion by waterflier** [buckyos/buckyos#54] [Announcements]
Hello everyone,  

We are thrilled to announce that the first public developer preview version, **0.3 Alpha1**, is now available! This release embodies our vision and showcases the core principles of a distributed operating system.  

For more details, please refer to the updated project [README.md](https://github.com/buckyos/buckyos/blob/main/README.md). We won’t repeat the introduction here. 

Thank you all for your continued support and valuable feedback. Join us on this journey! Feel free to open issues or submit pull requests. Let’s work together to build the next-generation distributed operating system!  

---

### Token Distribution Rules  

Following the **BuckyOS DAO** guidelines, we will release tokens to contributors based on the version release plan. (For detailed rules, visit: [http://dao.buckyos.org](http://dao.buckyos.org))  

While actual development progress diverged somewhat from the original plan (such is the gap between reality and ideals), we’ve essentially combined versions **0.2** and **0.3** into this release. A standalone release of 0.2, being purely kernel-focused, held limited value.  

To reflect these changes, I’ve updated the version roadmap in [BuckyOS 0.4 Alpha2 Plan.md](https://github.com/buckyos/buckyos/blob/main/doc/plan.md). Our target is to release the **Alpha version** of BuckyOS within the next 3-4 months. According to the roadmap:  
- **Alpha2** will be the first consumer-facing version.  
- **Alpha3** will introduce AI capabilities (OpenDAN integration).  

---

### Updated BuckyOS Release Plan  

**2024**  
- **0.1 Demo**: 2.5% (Done)  
- **0.2 PoC**: 2.5% (This Version)  
- **0.3 Alpha1**: 2.5% (This Version)  
- **0.4 Alpha2**: 2.5% (2024 Q4)  

**2025**  
- **0.5 Alpha3**: 5% (2025 Q1, First Public Test)  
- **0.6 Beta**: 5% (First public release)  
- **0.7 Release**: 2.5% (2025 Q3)  

---

### Alpha Phase Tasks  

Based on the updated plan, I’ve also reorganized the tasks for the Alpha phase. All **completed tasks** are eligible for token distribution in this round.  

This plan also outlines the primary goals for **0.4 Alpha2** (though based on experience, the actual workload may exceed what’s currently planned). However, this issue will focus solely on the **0.3 settlement**. Detailed discussions about the 0.4 roadmap will be addressed in a separate issue.  

---  
- Kernel Models (modules run before(support) system boot)
  - [x] node_daemon (A4 @waterflier)
    - [x] app & service loader (A4 @waterflier), implement formal permission management and container isolation
    - [ ] *node task execute system (A4), usually used for maintenance tasks; implement if unavoidable
  - [x] node_active (A4 @waterflier), System setup web pages and functions
  - [x] system config service(A2 @waterflier)
    - [ ] *Support etcd in more than 3 OOD clusters through scalable backends (A2)
    - [x] system-config lib (A2 @waterflier)
    - [x] rbac libs (A4 @waterflier), basic rbac usage and management
    - [ ] system config event support (A2 @waterflier), use websocket for config-change notification
    - [ ] *Integrate with cyfs-gateway's VPN service (A2 @waterflier)
  - [x] verify_hub service  (A2,@waterflier)
  - system status for system status monitoring
  - kRPC @waterflier
    - [x] rust kRPC libs (A4,@waterflier)
    - [x] typescript kRPC libs (A2,@waterflier)
      - [x] Improve user and device register  logical. (A2)
      - [x] Support Traditional user passwords log in and implement OAUTH compatible SDK (A2)
  - [x] *scheduler (A2 @waterflier), a key module to be implemented in the PoC version, automatically generating node_config and establishing an initial extensible framework
    - [x] *boot scheduler (A2,@waterflier), the first scheduler to be implemented, mainly for system initialization
    - [x] *scheduler template support(A1,@waterflier)
    - [ ] *Making scheduling logic for single OOD(A2)
    - [ ] *Implement the scheduling logic for multi-OOD(A2)
    - [ ] *When single OOD scale to multiple OOD, realize the scheduling logic with OP task(A4)
  - kLog, a reliable logging library, is the foundation for automatic fault diagnosis in the system.
    - [ ] *kLog lib (A4), defines the basic interfaces for kLog output and reliable behavior logic, can handle server downtime
    - [ ] *kLog server (S2), PoC version should implement a simple version to ensure reliability
  - pkg system
    - [x] Improve lib (S2,@waterflier) to facilitate use by other components
    - [ ] Integrate with the task system (A4,@glen0125)
- Kernel Services
  - [ ] *Task Manager (A4), providing a general stateful background task management service, supporting reliable execution of critical tasks
  - DFS
    - [ ] *glusterFS (A2) integrated with rbac
    - [ ] *glusterFS Support soft RAID: 4 hard disks can damage any hard disk without losing data （A4)
    - [ ] *glusterFS Support SSD Cache (A2)
    - DCFS (listed separately)
  - *dApp manager, the `apt` tool in BuckyOS, provides basic reliable pkg management capabilities for the system.
    - [ ] *basic API support (A4,@glen0125), source management, installed management, permission configuration, installer
    - [ ] *CLI tools (S4,@glen0125), command-line tools similar to apt based on basic API
    - [ ] *in-zone pkg repo service (S4,@glen0125), a stable repo service running within the zone
    - [ ] *Installer UI (A4), meets the minimum requirements for dApp permission control in the system
  - backup system (listed separately)
  - cyfs-gateway (listed separately)
  - kMQ message queue, supports custom event systems
- Frame Services
  - [ ] *smb-service (A2), integrated with ACL
  - [ ] k8s-service, integrated with ACL
  - [ ] Notify Manager
  - [ ] *msg_bus(S4), the user's system inbox, where all applications can send messages
  - [ ] *dApp Store
  - [ ] *Control panel (S8) Provide basic system management functions according to the requirements document (including related web pages), can be further divided.
- Default dApps
  - [x] Home Station App(A2,@waterflier),home page app for user, (Transform from https://github.com/filebrowser/filebrowser)
  - [x] System Control Panel App (A1,@waterflier)
    - [ ] *Account management (S4), mainly local DID account management, much logic can be reused from CYFS wallet
    - [ ] *Name management (S4), manage friendly names owned
    - [ ] *Zone management (S2), support relevant requirements of the first version
    - [ ] *Storage management (S2), pure display in the first version
  - [ ] CyberChat App (A4), BuckyOS control app modified based on the CYFS wallet app.
- Web3 bridge Services(Test: web3.buckyos.io,Officially web3.buckyos.org)
  - [x] Account management + signature service (A2), including signature history
  - [x] did resolution and name resolution (S2), mainly implemented in cyfs-gateway, this mainly handles formal online operations
  - [x] NAME manager (S4), allowing users to easily and freely own a $name.buckyos.org name, obtained during account registration,support d-dns
  - [x] rtcp network reply service (A2 @waterflier) (subscription management) 
    - [ ] rtcp network reply service Support billing and subscription (S4)
  - [ ] *WebUI (S4), a simple web page for users to manage their accounts and names annd subscriptions
  - [ ] *http backup server (S2), mainly functions from BuckyOS Backup Suite, simple online operation at first (with size restrictions), followed by subscription implementation
- [ ] *BuckyOS Backup Suite (independent product with separate points, additional rewards from the DMC fund), an independent cross-platform backup software, refer to its independent PRD.
  - [ ] Backup Service
  - [ ] UI
  - [ ] Backup basic libs
  - [ ] http DAV target client
  - [ ] http DAV target server
  - [ ] dmc target client
  - [ ] dmc target server
- [x] CYFS Gateway (A2,@waterflier) (would be a independent product after 0.4)
  - [x] tunnel framework (A2,@waterflier), A URL-based scalable Tunnel protocol framework, separating the business logic and protocol expansion of CYFS-Gateway
  - [x] rtcp protocol (A6,@waterflier),Based on TCP, a credible encrypted communication is realized based on DID, and OOD after the SN is transformed into NAT provides stable penetration access capabilities
  - [x] cyfs-dns (A1, @waterflier), supporting our name system and did-document system
  - [x] cyfs-warp (A4,@waterflier), A HTTP service that is base on Tunnel Framework can be regarded as Nginx-Lite
  - [ ] *cyfs-socks service(A3), Through the rules engine, the qualified traffic is forwarded to a specific Tunnel
  - [ ] *TAP Device-based VPN (A2), allowing needed services to work transparently with the main OOD in the same LAN
  - [ ] *cyfs protocl (A4), http extension only, support the chunk transfer logic used by backup system.
- CI/CD Publish Support
  - [x] Nightly CI/CD system (S4,@weiqiushi), based on Github Action
  - [x] deb(include arm) package builder (S2,@waterflier)
  - [ ] Virtual Machine Image builder (S2,@weiqiushi), based on Packer?
  - [ ] Rapid cloud development environment setup (A4), referencing Github's CodeSpace
  - [ ] Rapid independent CI/CD environment setup based on specific branches
  - [ ] Set up typical responsible test environments
- BuckyOS SDK 
  - [x] TypeScript SDK (A1,@waterflier)
    - [x] Auth Client (A2,@waterflier)
    - [x] kRPC Client (A1,@waterflier)
    - [ ] *File Share Client
    - [ ] Payment Gateway
---  

## Token Allocation Calculation Based on Contributions

Based on the statistics of work completed during the 0.2 & 0.3 release, we can calculate each individual's contribution proportion for this version according to the following rules:

- The workload of tasks is divided into A (Architect level) and S (Software Engineer level), with the number following each letter representing the required work weeks. The minimum time requirement for each task is 1 week.
- The score coefficient for A-level work is 3, and for S-level work, it is 2.
- If multiple people collaboratively complete a task (which is not encouraged), unless otherwise specified, the score is equally distributed among them.

Using these rules, we can calculate each individual's contribution for the 0.2 & 0.3 version.

Here is the token allocation score calculation based on the provided contributions:

#### @waterflier
1. **node_daemon (A4)**: 4 weeks * 3 = **12 points**
2. **app & service loader (A4)**: 4 weeks * 3 = **12 points**
3. **node_active (A4)**: 4 weeks * 3 = **12 points**
4. **system config service (A2)**: 2 weeks * 3 = **6 points**
5. **system-config lib (A2)**: 2 weeks * 3 = **6 points**
6. **rbac libs (A4)**: 4 weeks * 3 = **12 points**
7. **verify_hub service (A2)**: 2 weeks * 3 = **6 points**
8. **rust kRPC libs (A4)**: 4 weeks * 3 = **12 points**
9. **typescript kRPC libs (A2)**: 2 weeks * 3 = **6 points**
    - Improve user and device register logic (A2): 2 weeks * 3 = **6 points**
    - Support Traditional user passwords login (A2): 2 weeks * 3 = **6 points**
10. **scheduler (A2)**: 2 weeks * 3 = **6 points**
    - boot scheduler (A2): 2 weeks * 3 = **6 points**
    - scheduler template support (A1): 1 week * 3 = **3 points**
11. **pkg system (S2)**: 2 weeks * 2 = **4 points**
12. **Home Station App (A2)**: 2 weeks * 3 = **6 points**
13. **System Control Panel App (A1)**: 1 week * 3 = **3 points**
14. **rtcp network reply service (A2)**: 2 weeks * 3 = **6 points**
15. **CYFS Gateway (A2)**: 2 weeks * 3 = **6 points**
    - tunnel framework (A2): 2 weeks * 3 = **6 points**
    - rtcp protocol (A6)**: 6 weeks * 3 = **18 points**
    - cyfs-dns (A1): 1 week * 3 = **3 points**
    - cyfs-warp (A4): 4 weeks * 3 = **12 points**
16. **deb package builder (S2)**: 2 weeks * 2 = **4 points**
17. **TypeScript SDK (A1)**: 1 week * 3 = **3 points**
    - Auth Client (A2): 2 weeks * 3 = **6 points**
    - kRPC Client (A1): 1 week * 3 = **3 points**

Total for **@waterflier**: **159 points**

#### @weiqiushi
1. **Nightly CI/CD system (S4)**: 4 weeks * 2 = **8 points**

Total for **@weiqiushi**: **8 points**

### Summary
- **@waterflier**: **159 points**  The contribution is 95.2095%
- **@weiqiushi**: **8 points** The contribution is 4.7905%

In this version, many friends who are actively participating in DEMO have fallen into the dilemma that do not know how to start formal system writing, and it is free when not doing DEMO.I think this is difficult to start with everything ~ With the release of Alpha1, the logic of many "architectural documents" has become a code. I look forward to seeing more friends from the next version ~

---  
Let’s keep pushing forward together!

**Comment by glaxyfund** [comment]
Congratulations on the release of Alpha1!

The setup process was straightforward, and it's really easy to access and start using. The features provided are definitely sufficient for this stage and show great promise.

I understand that the data backup feature and the app store are not available yet, and I'm looking forward to seeing them in the next version. Additionally, I hope to see the "Buy another NAS for more space " feature soon as well.

To show our support for the project, we have initiated a new investment contract. The total amount is 35,000 USDT, with an exchange rate of 1 USDT: 85 BDT~ 

https://dao.buckyos.org/invest/3

Looking forward to seeing what comes next!

**Comment by waterflier** [comment]
Alpha1 is primarily a release aimed at developers, with most of the work happening behind the scenes. Once the foundation is stabilized, the speed of implementing new features will increase significantly.  

Thank you for your support!
